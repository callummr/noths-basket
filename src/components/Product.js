import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

const ProductCard = styled.article`
  @media (min-width: 415px) {
    width: 45%;
  }
  @media (min-width: 768px) {
    width: 30%;
  }
  width: 100%;
  margin-bottom: 20px;
  background-color: #fff;
  padding: 15px;
  position: relative;
  padding-bottom: 70px;
`

const Image = styled.img`
  width: 100%;
  margin-bottom: 10px;
`

const Title = styled.h1`
  color: #333333;
  font-size: 19px;
  margin-bottom: 15px;
  line-height: 1.2;
`

const Price = styled.span`
  color: green;
`

const Button = styled.button`
  position: absolute;
  height: 40px;
  background-color: #0a89af;
  bottom: 15px;
  left: 15px;
  right: 15px;
  line-height: 40px;
  text-align: center;
  text-decoration: none;
  font-size: 15px;
  color: #fff;
  border-radius: 5px;
  font-weight: 600;
  width: calc(100% - 30px);
  padding: 0;

  &:hover {
    opacity: 0.8;
  }
`

const Product = props => (
  <ProductCard>
    <Image
      src={props.imageUrl || 'http://via.placeholder.com/150x150'}
      alt={props.name}
    />
    <Title>{props.name}</Title>
    <Price>Â£{props.price.toFixed(2)}</Price>
    <Button onClick={() => props.onAddToBasket(props.id)}>Add to Basket</Button>
  </ProductCard>
)

Product.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  imageUrl: PropTypes.string,
  price: PropTypes.number.isRequired,
  onAddToBasket: PropTypes.func.isRequired,
}

export default Product
