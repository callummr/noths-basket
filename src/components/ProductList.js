import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import Product from './Product'

const Container = styled.section`
  @media (min-width: 768px) {
    width: calc(100% - 300px);
  }
  order: 0;
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  align-items: flex-start;
  width: 100%;
`

const ProductList = props => (
  <Container>
    {props.products.map(p => (
      <Product key={p.id} {...p} onAddToBasket={props.onAddToBasket} />
    ))}
  </Container>
)

ProductList.propTypes = {
  products: PropTypes.arrayOf(PropTypes.object).isRequired,
  onAddToBasket: PropTypes.func.isRequired,
}

export default ProductList
