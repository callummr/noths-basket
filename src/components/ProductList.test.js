import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import ProductList from './ProductList'
import Product from './Product'

const productListProps = {
  products: [
    { id: '001', name: 'test-name', price: 1.5, iamgeUrl: 'http://test.image' },
    { id: '002', name: 'test-name-2', price: 3.95 },
  ],
  onAddToBasket: jest.fn(),
}

describe('ProductList', () => {
  beforeEach(() => {
    productListProps.onAddToBasket.mockClear()
  })

  it('matches snapshot', () => {
    const instance = renderer.create(<ProductList {...productListProps} />)
    expect(instance.toJSON()).toMatchSnapshot()
  })

  it('renders as many Products as provided', () => {
    const wrapper = mount(<ProductList {...productListProps} />)
    expect(wrapper.find(Product).length).toBe(productListProps.products.length)
  })

  it('passes onAddToBasket down to Products', () => {
    const wrapper = mount(<ProductList {...productListProps} />)
    const aProduct = wrapper.find(Product).first()
    expect(aProduct.props()).toHaveProperty('onAddToBasket')
    aProduct.props().onAddToBasket()
    expect(productListProps.onAddToBasket).toHaveBeenCalled()
  })
})
