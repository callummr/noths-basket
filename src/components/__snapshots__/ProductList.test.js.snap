// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList matches snapshot 1`] = `
<section
  className="sc-EHOje jvCcfz"
>
  <article
    className="sc-bdVaJa bUUZdf"
  >
    <img
      alt="test-name"
      className="sc-bwzfXH qVTvB"
      src="http://via.placeholder.com/150x150"
    />
    <h1
      className="sc-htpNat lloKWr"
    >
      test-name
    </h1>
    <span
      className="sc-bxivhb jiHAbL"
    >
      £
      1.50
    </span>
    <button
      className="sc-ifAKCX cpSZLT"
      onClick={[Function]}
    >
      Add to Basket
    </button>
  </article>
  <article
    className="sc-bdVaJa bUUZdf"
  >
    <img
      alt="test-name-2"
      className="sc-bwzfXH qVTvB"
      src="http://via.placeholder.com/150x150"
    />
    <h1
      className="sc-htpNat lloKWr"
    >
      test-name-2
    </h1>
    <span
      className="sc-bxivhb jiHAbL"
    >
      £
      3.95
    </span>
    <button
      className="sc-ifAKCX cpSZLT"
      onClick={[Function]}
    >
      Add to Basket
    </button>
  </article>
</section>
`;
